mesonbuild/compilers/mixins/clike.py

```diff
@@ -1069,19 +1069,22 @@ def get_library_naming(self, env: 'Environment', libtype: LibType, strict: bool
 
     @staticmethod
     def _sort_shlibs_openbsd(libs: T.List[str]) -> T.List[str]:
+        def tuple_key(x: str) -> T.Tuple[int, ...]:
+            ver = x.rsplit('.so.', maxsplit=1)[1]
+            return tuple(int(i) for i in ver.split('.'))
+
         filtered: T.List[str] = []
         for lib in libs:
             # Validate file as a shared library of type libfoo.so.X.Y
             ret = lib.rsplit('.so.', maxsplit=1)
             if len(ret) != 2:
                 continue
             try:
-                float(ret[1])
+                tuple(int(i) for i in ret[1].split('.'))
             except ValueError:
                 continue
```

Comment:
I think we can avoid repeating the splitting algo
```python
try:
  if len(tuple_key(lib)) !=2:
    continue
except ValueError:
  continue    
```

I was hoping we could sort and filter without computing the key twice, but haven't found an easy way.

Reply:
What about:

```
            try:
                ver = tuple(int(i) for i in ret[1].split('.'))
            except ValueError:
                continue
            filtered.append((*ver, lib))
        return [f[3] for f in sorted(filtered, reverse=True)]
```

Reply:
Or, if you want to keep libname in alphabetical order:

```
            try:
                ver = tuple(-int(i) for i in ret[1].split('.'))
            except ValueError:
                continue
            filtered.append((*ver, lib))
        return [f[-1] for f in sorted(filtered)]
```

Reply:
@xclaesse the return value of tuple_key is a tuple of integers representing the version number, there is zero guarantee that it is length 2. Instead, we may get an uncaught IndexError that crashes meson. We could try to catch both IndexError and ValueError, but frankly I think we could do with less generic catching of generic error classes here, and it's not harmful to repeat this simple, simple logic.

@bruchar1,
Both of your solutions cause meson to crash fatally, with a `TypeError: '<' not supported between instances of 'str' and 'int'`. This is actually a pretty useful suggestion as it gave me an idea to enhance the test case a bit more. :p

Ultimately, I think there's no problem with the original solution and I'd rather not keep chasing questionable optimizations.

Reply:
@eli-schwartz my understanding is it must be exactly 2 numbers (https://github.com/mesonbuild/meson/issues/12195#issuecomment-1703298951) and float() was a very bad way of checking that. I'm not against being more flexible on the other hand.

Reply:
That could instead be:
```python
            try:
                tuple_key(lib)
            except (ValueError, IndexError):
                continue
```

Anyway, it's small nitpicking and it's merged already, nevermind.

Reply:
I'd still prefer not to error out with an internal bug in the event that a library file is misnamed. Part of our existing tests for the openbsd search logic already use invalid library names...